proposals/NNNN-implicit-member-chains.md

```diff
@@ -0,0 +1,250 @@
+# Extend implicit member syntax to cover chains of member references
+
+* Proposal: [SE-NNNN](NNNN-implicit-member-chains.md)
+* Authors: [Frederick Kellison-Linn](https://github.com/jumhyn)
+* Review Manager: TBD
+* Status: **Awaiting review**
+* Implementation: [apple/swift#31679](https://github.com/apple/swift/pull/31679)
+
+## Introduction
+
+When the type of an expression is implied by the context, Swift allows developers to use what is formally referred to as an "implicit member expression," sometimes referred to as "leading dot syntax":
+
+```swift
+class C {
+    static let zero = C(0)
+    var x: Int
+
+    init(_ x: Int) {
+        self.x = x
+    }
+}
+
+func f(_ c: C) {
+    print(c.x)
+}
+
+f(.zero) // prints '0'
+```
+
+This allows for the omission of repetitive type information in contexts where the type information is already obvious to the reader:
+```swift
+view.backgroundColor = .systemBackground
+```
+
+This proposal suggests the expansion of implicit member syntax to more complex expressions than just a single static member or function. Specifically, implicit member syntax would be allowed to cover chains of member references.
+
+## Motivation
+
+Today, attempting to use implicit member syntax with a chain of member references fails:
+
+```swift
+extension C {
+    var incremented: C {
+        return C(self.x + 1)
+    }
+}
+
+f(.zero.incremented) // Error: Type of expression is ambiguous without more context
+```
+
+This error breaks the mental model that many users likely have for implicit member syntax, which boils down to a simple lexical omission of the type name in contexts where the type is clear. I.e., users expect that writing:
+
+```swift
+let one: C = .zero.incremented
+```
+
+is just the same as writing
+
+```swift
+let one = C.zero.incremented
+```
+
+This issue arises in practice with any type that offers "modifier" methods that vend updated instances according to some rule. For example, `UIColor` offers the `withAlphaComponent(_:)` modifier for constructing new colors, which cannot be used with implicit member syntax:
+
+```swift
+let milky: UIColor = .white.withAlphaComponent(0.5) // error
+```
+
+## Proposed solution
+
+Improve in order to be able to handle multiple chained member accesses. The type of the resulting expression would be constrained to match the contextual type. Under this proposal, all of the following would successfully typecheck:
```

Comment:
Improve **???** in order to be able to handle multiple chained member accesses.

Reply:
```suggestion
Improve implicit member expressions to handle multiple chained member accesses. The type of the resulting expression would be constrained to match the contextual type. Under this proposal, all of the following would successfully typecheck:
```